File: packages/arcis-python/arcis/middleware/csrf.py
"""
Arcis Middleware - CSRF Protection

Cross-Site Request Forgery protection using double-submit cookie pattern:
1. Server sets a CSRF token in a cookie
2. Client must send the same token in a header or form field
3. Middleware rejects requests where cookie token != header/field token

This works because an attacker's cross-origin form can include the cookie
automatically, but cannot read it (same-origin policy) to set the header.
"""

import hmac
import os
import secrets
from dataclasses import dataclass, field
from typing import Callable, List, Optional, Sequence


DEFAULT_COOKIE_NAME = "_csrf"
DEFAULT_HEADER_NAME = "x-csrf-token"
DEFAULT_FIELD_NAME = "_csrf"
DEFAULT_TOKEN_LENGTH = 32
DEFAULT_PROTECTED_METHODS = ("POST", "PUT", "PATCH", "DELETE")
HOST_PREFIX = "__Host-"


def generate_csrf_token(length: int = DEFAULT_TOKEN_LENGTH) -> str:
    """
    Generate a cryptographically random CSRF token.

    Args:
        length: Byte length (output is hex, so 2x chars). Default: 32

    Returns:
        Hex-encoded random token (64 chars by default)

    Example:
        token = generate_csrf_token()  # 64 hex chars
    """
    return secrets.token_hex(length)


def validate_csrf_token(cookie_token: str, request_token: str) -> bool:
    """
    Validate that two CSRF tokens match using constant-time comparison.

    Args:
        cookie_token: Token from the cookie
        request_token: Token from the header or form field

    Returns:
        True if tokens match
    """
    if not cookie_token or not request_token:
        return False
    return hmac.compare_digest(cookie_token, request_token)


@dataclass
class CsrfCookieOptions:
    """Cookie options for the CSRF token."""

    path: str = "/"
    http_only: bool = False  # Must be readable by client JS
    secure: Optional[bool] = None  # None = auto (production detection)
    same_site: str = "Lax"
    domain: Optional[str] = None


@dataclass
class CsrfOptions:
    """CSRF protection configuration."""

    cookie_name: str = DEFAULT_COOKIE_NAME
    header_name: str = DEFAULT_HEADER_NAME
    field_name: str = DEFAULT_FIELD_NAME
    token_length: int = DEFAULT_TOKEN_LENGTH
    protected_methods: Sequence[str] = DEFAULT_PROTECTED_METHODS
    exclude_paths: List[str] = field(default_factory=list)
    cookie: CsrfCookieOptions = field(default_factory=CsrfCookieOptions)


class CsrfProtection:
    """
    CSRF protection using double-submit cookie pattern.

    For safe methods (GET, HEAD, OPTIONS), sets a CSRF token cookie.
    For unsafe methods (POST, PUT, PATCH, DELETE), validates the token.

    Example (Flask):
        csrf = CsrfProtection()

        @app.before_request
        def check_csrf():
            error = csrf.flask_before_request()
            if error:
                return error

        @app.after_request
        def set_csrf_cookie(response):
            return csrf.flask_after_request(response)

    Example (client-side):
        token = document.cookie.match(/_csrf=([^;]+)/)?.[1]
        fetch('/api/data', {
            method: 'POST',
            headers: { 'X-CSRF-Token': token },
            credentials: 'same-origin'
        })
    """

    def __init__(
        self,
        cookie_name: str = DEFAULT_COOKIE_NAME,
        header_name: str = DEFAULT_HEADER_NAME,
        field_name: str = DEFAULT_FIELD_NAME,
        token_length: int = DEFAULT_TOKEN_LENGTH,
        protected_methods: Optional[Sequence[str]] = None,
        exclude_paths: Optional[List[str]] = None,
        cookie: Optional[CsrfCookieOptions] = None,
        on_error: Optional[Callable] = None,
        use_host_prefix: bool = False,
        skip_csrf: Optional[Callable] = None,
    ):
        # __Host- prefix: forces browser to enforce Secure + no Domain + Path=/
        self.cookie_name = f"{HOST_PREFIX}{cookie_name}" if use_host_prefix else cookie_name
        self.use_host_prefix = use_host_prefix
        self.skip_csrf = skip_csrf
        self.header_name = header_name
        self.field_name = field_name
        self.token_length = token_length
        self.protected_methods = set(
            m.upper() for m in (protected_methods or DEFAULT_PROTECTED_METHODS)
        )
        self.exclude_paths = exclude_paths or []
        self.cookie_opts = cookie or CsrfCookieOptions()
        self.on_error = on_error

    def _is_excluded(self, path: str) -> bool:
        """Check if a path is excluded from CSRF protection."""
        for excluded in self.exclude_paths:
            if path == excluded or path.startswith(excluded + "/"):
                return True
        return False

    def _build_cookie_header(self, token: str) -> str:
        """Build a Set-Cookie header value for the CSRF token."""
        parts = [f"{self.cookie_name}={token}"]
        parts.append(f"Path={self.cookie_opts.path}")
        if self.cookie_opts.http_only:
            parts.append("HttpOnly")

        secure = self.cookie_opts.secure
        if secure is None:
            secure = os.environ.get("FLASK_ENV") != "development"
        if secure:
            parts.append("Secure")

        parts.append(f"SameSite={self.cookie_opts.same_site}")
        if self.cookie_opts.domain:
            parts.append(f"Domain={self.cookie_opts.domain}")
        return "; ".join(parts)

    # ── Flask Integration ──────────────────────────────────────────────────

    def flask_before_request(self):
        """
        Flask before_request handler. Returns None on success, or a
        (response_body, status_code) tuple on CSRF failure.

        Example:
            @app.before_request
            def check_csrf():
                error = csrf.flask_before_request()
                if error:
                    return error
        """
        from flask import request

        # Per-request skip callback (API keys, signed webhooks, etc.)
        if self.skip_csrf and self.skip_csrf(request):
            return None

        if self._is_excluded(request.path):
            return None

        method = request.method.upper()
        if method not in self.protected_methods:
            return None

        cookie_token = request.cookies.get(self.cookie_name)
        if not cookie_token:
            return self._flask_error()

        request_token = self._get_flask_request_token(request)
        if not request_token:
            return self._flask_error()

        if not validate_csrf_token(cookie_token, request_token):
            return self._flask_error()

        return None

    def flask_after_request(self, response):
        """
        Flask after_request handler. Sets CSRF cookie on safe method responses
        if no cookie exists yet.

        Example:
            @app.after_request
            def set_csrf(response):
                return csrf.flask_after_request(response)
        """
        from flask import request

        method = request.method.upper()
        if method in self.protected_methods:
            return response

        existing = request.cookies.get(self.cookie_name)
        if not existing:
            token = generate_csrf_token(self.token_length)
            response.headers.add("Set-Cookie", self._build_cookie_header(token))

        return response

    def _get_flask_request_token(self, request) -> Optional[str]:
        """Extract CSRF token from Flask request (header or form body only)."""
        # Header
        header_token = request.headers.get(self.header_name)
        if header_token:
            return header_token

        # Form field
        form_token = request.form.get(self.field_name)
        if form_token:
            return form_token

        # JSON body
        if request.is_json:
            try:
                json_data = request.get_json(silent=True)
                if json_data and isinstance(json_data, dict):
                    token = json_data.get(self.field_name)
                    if token and isinstance(token, str):
                        return token
            except Exception:
                pass

        # SECURITY: Query string intentionally not supported — tokens in URLs leak
        # to server logs, Referer headers, browser history, and CDN/proxy logs.

        return None

    def _flask_error(self):
        """Return Flask CSRF error response."""
        if self.on_error:
            from flask import request
            return self.on_error(request)
        return {
            "error": "CSRF token validation failed",
            "message": "Invalid or missing CSRF token. Include the token "
                       "from the cookie in the X-CSRF-Token header.",
        }, 403

    # ── Generic / Framework-Agnostic ───────────────────────────────────────

    def check(
        self,
        method: str,
        path: str,
        cookie_token: Optional[str],
        request_token: Optional[str],
    ) -> bool:
        """
        Generic CSRF validation (framework-agnostic).

        Args:
            method: HTTP method (GET, POST, etc.)
            path: Request path
            cookie_token: Token from the cookie
            request_token: Token from header or body

        Returns:
            True if request is valid (safe method or valid token)
        """
        if self._is_excluded(path):
            return True

        if method.upper() not in self.protected_methods:
            return True

        if not cookie_token or not request_token:
            return False

        return validate_csrf_token(cookie_token, request_token)

    def generate_token(self) -> str:
        """Generate a new CSRF token."""
        return generate_csrf_token(self.token_length)


def create_csrf(
    cookie_name: str = DEFAULT_COOKIE_NAME,
    header_name: str = DEFAULT_HEADER_NAME,
    field_name: str = DEFAULT_FIELD_NAME,
    token_length: int = DEFAULT_TOKEN_LENGTH,
    protected_methods: Optional[Sequence[str]] = None,
    exclude_paths: Optional[List[str]] = None,
    cookie: Optional[CsrfCookieOptions] = None,
    on_error: Optional[Callable] = None,
    use_host_prefix: bool = False,
    skip_csrf: Optional[Callable] = None,
) -> CsrfProtection:
    """
    Create a CSRF protection handler.

    Example:
        csrf = create_csrf(exclude_paths=['/api/webhooks'])
    """
    return CsrfProtection(
        cookie_name=cookie_name,
        header_name=header_name,
        field_name=field_name,
        token_length=token_length,
        protected_methods=protected_methods,
        exclude_paths=exclude_paths,
        cookie=cookie,
        on_error=on_error,
        use_host_prefix=use_host_prefix,
        skip_csrf=skip_csrf,
    )

File: packages/arcis-node/src/middleware/csrf.ts
/**
 * @module @arcis/node/middleware/csrf
 * CSRF (Cross-Site Request Forgery) protection middleware
 *
 * Implements the double-submit cookie pattern:
 * 1. Server sets a CSRF token in a cookie
 * 2. Client must send the same token in a header or form field
 * 3. Middleware rejects requests where cookie token !== header/field token
 *
 * This works because an attacker's cross-origin form submission will include
 * the cookie automatically, but cannot read it (same-origin policy) to set
 * the matching header.
 */

import { randomBytes, timingSafeEqual } from 'crypto';
import type { Request, Response, NextFunction, RequestHandler } from 'express';

/** CSRF protection configuration */
export interface CsrfOptions {
  /** Cookie name for the CSRF token. Default: '_csrf' */
  cookieName?: string;
  /** Header name to check for the token. Default: 'x-csrf-token' */
  headerName?: string;
  /** Form field name to check for the token. Default: '_csrf' */
  fieldName?: string;
  /** Token byte length (hex-encoded = 2x chars). Default: 32 */
  tokenLength?: number;
  /** HTTP methods to protect. Default: ['POST', 'PUT', 'PATCH', 'DELETE'] */
  protectedMethods?: string[];
  /** Paths to exclude from CSRF checks (e.g., webhook endpoints) */
  excludePaths?: string[];
  /**
   * Per-request skip function. If it returns true, CSRF check is skipped
   * for that request. Useful for API key auth or signed webhooks.
   *
   * @example
   * skipCsrf: (req) => Boolean(req.headers['x-api-key'])
   */
  skipCsrf?: (req: Request) => boolean;
  /**
   * Use the __Host- cookie prefix for stronger cookie security.
   * When enabled, the browser enforces: Secure=true, no Domain, Path=/.
   * This prevents CSRF cookie theft across subdomains.
   * Default: false
   */
  useHostPrefix?: boolean;
  /** Cookie options */
  cookie?: {
    /** Cookie path. Default: '/' */
    path?: string;
    /** HttpOnly — set false so client JS can read it for headers. Default: false */
    httpOnly?: boolean;
    /** Secure flag (HTTPS only). Default: true in production */
    secure?: boolean;
    /** SameSite attribute. Default: 'Lax' */
    sameSite?: 'Strict' | 'Lax' | 'None';
    /** Cookie domain */
    domain?: string;
  };
  /** Custom error handler when CSRF validation fails */
  onError?: (req: Request, res: Response, next: NextFunction) => void;
  /**
   * Rotate the CSRF token after each successful validation on a protected
   * method. Defends against token-fixation attacks where an attacker plants
   * a known token before authentication. Default: false.
   *
   * When enabled, every successful POST/PUT/PATCH/DELETE causes the server to
   * issue a fresh token via Set-Cookie — the client must re-read the cookie
   * before the next mutating request.
   */
  rotateOnUse?: boolean;
}

const DEFAULTS = {
  cookieName: '_csrf',
  headerName: 'x-csrf-token',
  fieldName: '_csrf',
  tokenLength: 32,
  protectedMethods: ['POST', 'PUT', 'PATCH', 'DELETE'],
} as const;

/**
 * Generate a cryptographically random CSRF token.
 *
 * @param length - Byte length (output is hex, so 2x chars). Default: 32
 * @returns Hex-encoded random token
 *
 * @example
 * const token = generateCsrfToken(); // 64 hex chars
 */
export function generateCsrfToken(length: number = 32): string {
  return randomBytes(length).toString('hex');
}

/**
 * Validate that two CSRF tokens match using constant-time comparison.
 *
 * @param cookieToken - Token from the cookie
 * @param requestToken - Token from the header or form field
 * @returns true if tokens match
 */
export function validateCsrfToken(cookieToken: string, requestToken: string): boolean {
  if (!cookieToken || !requestToken) return false;
  if (cookieToken.length !== requestToken.length) return false;

  // SECURITY: Use Node.js built-in constant-time comparison to prevent timing attacks
  const a = Buffer.from(cookieToken);
  const b = Buffer.from(requestToken);
  return timingSafeEqual(a, b);
}

/**
 * Extract the CSRF token from a request (checks header, then body field, then query).
 */
function getRequestToken(req: Request, headerName: string, fieldName: string): string | undefined {
  // 1. Check header (most common for SPAs)
  const headerToken = req.headers[headerName.toLowerCase()];
  if (typeof headerToken === 'string' && headerToken) return headerToken;

  // 2. Check body field (form submissions)
  if (req.body && typeof req.body === 'object' && fieldName in req.body) {
    const bodyToken = req.body[fieldName];
    if (typeof bodyToken === 'string' && bodyToken) return bodyToken;
  }

  // SECURITY: Query string intentionally not supported — tokens in URLs leak
  // to server logs, Referer headers, browser history, and CDN/proxy logs.

  return undefined;
}

/**
 * Create CSRF protection middleware using double-submit cookie pattern.
 *
 * For safe methods (GET, HEAD, OPTIONS), sets a CSRF token cookie if not present.
 * For unsafe methods (POST, PUT, PATCH, DELETE), validates the token.
 *
 * @param options - CSRF configuration
 * @returns Express middleware
 *
 * @example
 * // Basic usage
 * app.use(csrfProtection());
 *
 * @example
 * // Exclude webhook paths
 * app.use(csrfProtection({
 *   excludePaths: ['/api/webhooks/stripe', '/api/webhooks/github']
 * }));
 *
 * @example
 * // Client-side: read cookie + set header
 * const token = document.cookie.match(/_csrf=([^;]+)/)?.[1];
 * fetch('/api/data', {
 *   method: 'POST',
 *   headers: { 'X-CSRF-Token': token },
 *   credentials: 'same-origin'
 * });
 */
export function csrfProtection(options: CsrfOptions = {}): RequestHandler {
  const baseCookieName = options.cookieName ?? DEFAULTS.cookieName;
  // __Host- prefix: forces browser to enforce Secure + no Domain + Path=/
  const cookieName = options.useHostPrefix ? `__Host-${baseCookieName}` : baseCookieName;
  const headerName = options.headerName ?? DEFAULTS.headerName;
  const fieldName = options.fieldName ?? DEFAULTS.fieldName;
  const tokenLength = options.tokenLength ?? DEFAULTS.tokenLength;
  const protectedMethods = options.protectedMethods ?? [...DEFAULTS.protectedMethods];
  const excludePaths = options.excludePaths ?? [];
  const skipCsrf = options.skipCsrf;

  const isProduction = process.env.NODE_ENV === 'production';
  const cookieOpts = {
    path: options.cookie?.path ?? '/',
    httpOnly: options.cookie?.httpOnly ?? false, // Must be readable by client JS
    secure: options.cookie?.secure ?? isProduction,
    sameSite: options.cookie?.sameSite ?? 'Lax',
    domain: options.cookie?.domain,
  };

  const defaultOnError = (req: Request, res: Response, _next: NextFunction) => {
    // Telemetry attribution: dashboard groups CSRF denials under vector=csrf.
    req.__arcis = {
      vector: 'csrf',
      rule: 'csrf/token-mismatch',
      severity: 'high',
      reason: 'CSRF token missing or invalid',
      decision: 'deny',
    };
    res.status(403).json({
      error: 'CSRF token validation failed',
      message: 'Invalid or missing CSRF token. Include the token from the cookie in the X-CSRF-Token header.',
    });
  };

  const onError = options.onError ?? defaultOnError;

  // Normalize protected methods to uppercase
  const protectedSet = new Set(protectedMethods.map(m => m.toUpperCase()));

  return (req: Request, res: Response, next: NextFunction) => {
    const method = req.method.toUpperCase();

    // Per-request skip callback (API keys, signed webhooks, etc.)
    if (skipCsrf && skipCsrf(req)) {
      return next();
    }

    // Check if path is excluded
    const requestPath = req.path || req.url;
    if (excludePaths.some(p => requestPath === p || requestPath.startsWith(p + '/'))) {
      return next();
    }

    // Expose token generation on the request for templates/views
    (req as unknown as Record<string, unknown>).csrfToken = () => {
      const existing = getCookieValue(req, cookieName);
      if (existing) return existing;

      const token = generateCsrfToken(tokenLength);
      setCsrfCookie(res, cookieName, token, cookieOpts);
      return token;
    };

    // For safe methods — ensure a CSRF cookie exists
    if (!protectedSet.has(method)) {
      const existing = getCookieValue(req, cookieName);
      if (!existing) {
        const token = generateCsrfToken(tokenLength);
        setCsrfCookie(res, cookieName, token, cookieOpts);
      }
      return next();
    }

    // For protected methods — validate the token
    const cookieToken = getCookieValue(req, cookieName);
    if (!cookieToken) {
      return onError(req, res, next);
    }

    const requestToken = getRequestToken(req, headerName, fieldName);
    if (!requestToken) {
      return onError(req, res, next);
    }

    if (!validateCsrfToken(cookieToken, requestToken)) {
      return onError(req, res, next);
    }

    // Optional token rotation on successful validation
    if (options.rotateOnUse) {
      const freshToken = generateCsrfToken(tokenLength);
      setCsrfCookie(res, cookieName, freshToken, cookieOpts);
    }

    next();
  };
}

/**
 * Read a cookie value from the request.
 */
function getCookieValue(req: Request, name: string): string | undefined {
  // Express parses cookies if cookie-parser is used
  if (req.cookies && typeof req.cookies === 'object' && name in req.cookies) {
    return req.cookies[name];
  }

  // Fallback: parse from raw Cookie header
  const cookieHeader = req.headers.cookie;
  if (!cookieHeader) return undefined;

  const match = cookieHeader.match(new RegExp(`(?:^|;\\s*)${escapeRegex(name)}=([^;]*)`));
  return match ? decodeURIComponent(match[1]) : undefined;
}

/**
 * Set the CSRF token cookie on the response.
 */
function setCsrfCookie(
  res: Response,
  name: string,
  token: string,
  opts: { path: string; httpOnly: boolean; secure: boolean; sameSite: string; domain?: string }
): void {
  const parts = [`${name}=${token}`];
  parts.push(`Path=${opts.path}`);
  if (opts.httpOnly) parts.push('HttpOnly');
  if (opts.secure) parts.push('Secure');
  parts.push(`SameSite=${opts.sameSite}`);
  if (opts.domain) parts.push(`Domain=${opts.domain}`);

  // Accumulate Set-Cookie headers to avoid overwriting cookies set by other middleware
  const newCookie = parts.join('; ');
  const existing = res.getHeader('Set-Cookie');
  if (existing === undefined) {
    res.setHeader('Set-Cookie', newCookie);
  } else if (Array.isArray(existing)) {
    res.setHeader('Set-Cookie', [...existing, newCookie]);
  } else {
    res.setHeader('Set-Cookie', [existing as string, newCookie]);
  }
}

/**
 * Escape special regex characters in a string.
 */
function escapeRegex(str: string): string {
  return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

/** Alias for csrfProtection */
export const createCsrf = csrfProtection;
